tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -343,6 +343,210 @@ def _calc_q(n, m, sign, k=1):
+
+    if dividend_rates is None:
+      dividend_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='dividend_rates')
+
+    if cost_of_carries is not None:
+      cost_of_carries = tf.convert_to_tensor(
+          cost_of_carries, dtype=dtype, name='cost_of_carries')
+    else:
+      cost_of_carries = discount_rates - dividend_rates
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp(cost_of_carries * expiries)
+
+    # If cost of carry is greater than or equal to discount rate, then use Black-Scholes price
+    american_prices = tf.where(cost_of_carries >= discount_rates and is_call_options,
+            vanilla_prices.option_price(
+                volatilities=volatilities,
+                strikes=strikes,
+                expiries=expiries,
+                forwards=forwards,
+                discount_rates=discount_rates,
+                cost_of_carries=cost_of_carries,
+                is_call_options=is_call_options),
+            tf.where(is_call_options,
+                _call_1993(forwards, strikes, expiries, discount_rates, cost_of_carries, volatilities),
+                _call_1993(strikes, forwards, expiries, discount_rates - cost_of_carries, -cost_of_carries, volatilities)))
+
+    return american_prices
+
+
+def _call_1993(S, X, T, r, b, sigma):
+  beta = (0.5 - b / sigma**2) + tf.math.sqrt((b / sigma**2 - 0.5)**2 + 2 * r / sigma**2)
+  alpha = (_boundary_1993(X, T, r, b, sigma, beta) - X) * (_boundary_1993(X, T, r, b, sigma, beta))**(-beta)
+
+  return tf.where(S >= _boundary_1993(X, T, r, b, sigma, beta),
+          S - X,
+          alpha * S**beta -
+          alpha * _phi_1993(S, T, beta, _boundary_1993(X, T, r, b, sigma, beta), _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) +
+          _phi_1993(S, T, 1, _boundary_1993(X, T, r, b, sigma, beta), _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) -
+          _phi_1993(S, T, 1, X, _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) - 
+          X * _phi_1993(S, T, 0, _boundary_1993(X, T, r, b, sigma, beta), _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma) +
+          X * _phi_1993(S, T, 0, X, _boundary_1993(X, T, r, b, sigma, beta), r, b, sigma))
+
+
+def _phi_1993(S, T, gamma, H, I, r, b, sigma):
+  kappa = (2 * b) / sigma**2 + (2 * gamma - 1)
+  d = -(tf.math.log(S/H) + (b + (gamma - 0.5) * sigma**2) * T) / (sigma * tf.math.sqrt(T))
+  lambd = (-r + gamma * b + 0.5 * gamma * (gamma - 1) * sigma**2) * T
+
+  return tf.math.exp(lambd) * S**gamma * (_ncdf(d) - (I/S)**kappa * 
+          _ncdf(d - (2 * tf.math.log(I/S)) / (sigma * tf.math.sqrt(T))))
+
+
+def _boundary_1993(X, T, r, b, sigma, beta):
+  """Computes the boundary or trigger price for the Bjerksund-Stensland 1993 algorithm."""
+  B0 = tf.maximum(X, (r / (r - b)) * X)
```

Comment:
Please avoid using capital letters in variable names, `b0` instead (according to the Python style guide). Here and elsewhere

Reply:
also, `tf.math.maximum` for consistency

Reply:
So should I change `X` and `T` to `x` and `t`?  That would make them inconsistent with the paper and there are equations in the paper that have both capital and lower case of the same letter being used in the same equation, especially in the 2002 algorithm.

Reply:
What about changing `X` to `upper_x` and `x` to `lower_x` or something like that?  Then someone putting the code and the paper side-by-side could still compare the two, but this would also adhere to the Python style guide?  Is that an acceptable solution?